### DISCUSSION TITLE
Main Map Node Display

### DISCUSSION THREAD

**Discussion by Warren-Whitfield** [pelicanmapping/osgearth#2555] [Support Q&A] [answered]
Is there a way to hide/show the main map node while still displaying all other graphics like vectors etc.? I'm assuming it has something to do with setting the nodeMask, but I'm not exactly sure?

**Comment by gwaldron** [comment]
You can try calling
`mapnode->getTerrainEngine()->setNodeMask(mask);`
Where mask==0 to hide, mask==~0 to show.

**Comment by Warren-Whitfield** [comment]
Ok, so m_poMapNode->getTerrainEngine()->getNode()->setNodeMask(~0 or 0) works to show/hide the map. 

The issue I now have is that I loose the getGeoPointUnderMouse() functionality on which many of my user actions are based, as well as the MouseCoordsTool() readout. However, any graphics that have already been drawn before hiding the map are retained, and incoming plots and tracks with georef positions are also still rendered ok.

So your suggestion is almost what I want, but with a few shortcomings... 
I guess the question now is how to get getGeoPointUnderMouse() and friends working when the map is not displayed? (if that is even possible)

**Comment by gwaldron** [comment]
Makes sense.
I added a utility class called `ToggleVisibleCullCallback`. This was originally used internally by the `VisibleLayer` class, but I figured it makes sense to expose it. You can add this callback to the `TerrainEngineNode` and the use `setVisible()` to toggle it. It will continue to page data in and out, and it will work with function like `getGeoPointUnderMouse`. Let me know if this works for you.

**Comment by Warren-Whitfield** [comment]
Glenn, I'm afraid I'm stuck here... _m_poMapNode->getTerrainEngine()->_ 
Can you show me an example of what I should be doing please?

**Reply by gwaldron** [reply, marked as answer]
```c++
#include <osgEarth/CullingUtils>
...
// setup:
m_callback = new osgEarth::Util::ToggleVisibleCullCallback();
m_poMapNode->getTerrainEngine()->getNode()->addCullCallback(m_callback );
...
// toggle:
m_callback ->setVisible(...);
```

**Comment by Warren-Whitfield** [comment]
Cool thx, will try it in the morning.

**Comment by Warren-Whitfield** [comment]
So in the version I'm using there is no ToggleVisibleCullCallback class/struct. I'm assuming it should be in the osgEarth/CullingUtils header, but it isn't. 
I'm using osgEarth from vcpkg.

**Reply by gwaldron** [reply]
Right, I just pushed it to the master branch a couple days ago.
You can just copy the class to your own code and use it there until you are ready to upgrade.

**Comment by Warren-Whitfield** [comment]
Dude!, Your'e the man! It works perfectly.
